ports/espressif/common-hal/busio/SPI.c

```diff
@@ -213,47 +214,62 @@ bool common_hal_busio_spi_transfer(busio_spi_obj_t *self,
 
     // Round to nearest whole set of bits
     int bits_to_send = len * 8 / self->bits * self->bits;
 
     if (len <= 4) {
+        memset(&transactions[0], 0, sizeof(spi_transaction_t));
         if (data_out != NULL) {
-            memcpy(&transaction.tx_data, data_out, len);
+            memcpy(&transactions[0].tx_data, data_out, len);
         }
 
-        transaction.flags = SPI_TRANS_USE_TXDATA | SPI_TRANS_USE_RXDATA;
-        transaction.length = bits_to_send;
-        spi_device_transmit(spi_handle[self->host_id], &transaction);
+        transactions[0].flags = SPI_TRANS_USE_TXDATA | SPI_TRANS_USE_RXDATA;
+        transactions[0].length = bits_to_send;
+        spi_device_transmit(spi_handle[self->host_id], &transactions[0]);
 
         if (data_in != NULL) {
-            memcpy(data_in, &transaction.rx_data, len);
+            memcpy(data_in, &transactions[0].rx_data, len);
         }
     } else {
         int offset = 0;
         int bits_remaining = bits_to_send;
+        int cur_trans = 0;
 
         while (bits_remaining && !mp_hal_is_interrupted()) {
-            memset(&transaction, 0, sizeof(transaction));
 
-            transaction.length =
-                bits_remaining > SPI_MAX_DMA_BITS ? SPI_MAX_DMA_BITS : bits_remaining;
+            cur_trans = 0;
+            while (bits_remaining && (cur_trans != MAX_SPI_TRANSACTIONS)) {
+                memset(&transactions[cur_trans], 0, sizeof(spi_transaction_t));
 
-            if (data_out != NULL) {
-                transaction.tx_buffer = data_out + offset;
-            }
-            if (data_in != NULL) {
-                transaction.rx_buffer = data_in + offset;
-            }
+                transactions[cur_trans].length =
+                    bits_remaining > SPI_MAX_DMA_BITS ? SPI_MAX_DMA_BITS : bits_remaining;
+
+                if (data_out != NULL) {
+                    transactions[cur_trans].tx_buffer = data_out + offset;
+                }
+                if (data_in != NULL) {
+                    transactions[cur_trans].rx_buffer = data_in + offset;
+                }
+
+                bits_remaining -= transactions[cur_trans].length;
 
-            spi_device_transmit(spi_handle[self->host_id], &transaction);
-            bits_remaining -= transaction.length;
+                // doesn't need ceil(); loop ends when bits_remaining is 0
+                offset += transactions[cur_trans].length / 8;
+                cur_trans++;
+            }
 
-            // doesn't need ceil(); loop ends when bits_remaining is 0
-            offset += transaction.length / 8;
+            for (int i = 0; i < cur_trans; i++) {
+                spi_device_queue_trans(spi_handle[self->host_id], &transactions[i], portMAX_DELAY);
+            }
 
             RUN_BACKGROUND_TASKS;
+
+            spi_transaction_t *rtrans;
+            for (int x = 0; x < cur_trans; x++) {
+                spi_device_get_trans_result(spi_handle[self->host_id], &rtrans, portMAX_DELAY);
```

Comment:
Does this block until the transaction is done? We may want to move `RUN_BACKGROUND_TASKS` here so that it can run after each transaction completes.

Reply:
Moved it there, and it did not cause any noticeable slowdowns.

I tested everything at 100Khz and never had a USB timeout and Ctrl-C is almost instant (even 4KB is 1/3 second at that speed).

Reply:
`spi_device_get_trans_result` is blocking until the transaction finishes (when the delay is set to `portMAX_DELAY`). So we ensure all (up to) 10 transactions return before we move on.

The old code used `spi_device_transmit` which combines `spi_device_queue_trans` and `spi_device_get_trans_result` into one function call, resulting in the same functionality (but with 1 transaction instead of 10).

In the new way up to 10 (as needed) transactions are queued up and then the code waits until it gets 10 results before continuing the loop. `bits_remaining` acts as a count while we set up the transactions.

Reply:
Ok, that's what I thought. Please move `RUN_BACKGROUND_TASKS;` into this loop because it is the one that will take a while.

Reply:
Sorry, totally blanked on which loop would actually be the slow one. Moved the check and tested down to 50Khz and nothing timed out on me.